### ISSUE TITLE
Faint humming (buzz) sound while running a game (HDMI RetroArch & RetroPie)

### ISSUE THREAD

**Issue by nanoant** [atari800/atari800#13]
I have recently managed to run Atari800 on my NVIDIA Shield TV with 2 PRs merged to libretro version https://github.com/libretro/libretro-atari800/pull/9 https://github.com/libretro/libretro-atari800/pull/10 

Everything works great except rather faint humming (buzz) sound emitted by atari800 emulator when running.  Sound goes via HDMI output, so it is not analog/hw problem, also no other emulators or Android apps exhibit this problem on my box.

I have also found another user having exactly same problem but in RetroPie, completely different hardware and platform, so I suppose it makes sense to blame atari800 emulator.

https://retropie.org.uk/forum/topic/6927/atari800-5200-emulation-faint-humming-sound-while-running-a-game

Is it some well known problem? Is there any workaround? (except rather ugly workaround removing the humming sound via equalizer or some DSP filter).

**Comment by knik0** [comment]
I have a guess, normally when running, atai800 puts full range DC voltage on the audio output, it might cause a buzz in some amplifiers.

If you tested recent build of atari800 with -volume option, something like atari800 -volume 10 it could reduce the buzz.

**Comment by nanoant** [comment]
## Update

I think I know the reason, I have dumped `atari800` generated sound into WAV and I found out that silence is not in fact a silence `uint8` == `128` (signed zero), but `33` instead !!! (-75%), moreover it seems that the wave is not zero aligned, see sshot below.

<img width="258" alt="atari800wav" src="https://user-images.githubusercontent.com/103067/32418043-ed4bec44-c262-11e7-99f7-132c58b9bead.png">

So altogether it looks like a bug, and I can expect that depending on the device, playing negative constant value instead zero for silence can produce so unpleasant (humming, buzz) effects.

**Comment by knik0** [comment]
Ah, you just confirmed what I said earlier.
Apparently you have piss-poor audio amp in your hdmi receiver.

**Comment by nanoant** [comment]
@knik0 Yeah that might be, it is just high-tier Samsung TV that I got as a present 😉 But I anyway I don't see why the amp is to blame if the audio signal is a mess, -75% constant DC for a silence is definitely not right.
Is physical Atari generating such a bizarre signal too?

**Comment by knik0** [comment]
I guess it was done this way to maximize output volume.
AFAIK the real Atari was producing sound it a similar way. It is sometimes used to play digitized samples.
I guess it can be fixed after reducing sound volume, I mean output range 0.0 to 1.0 instead of -1.0 to 1.0

**Comment by joysfera** [comment]
Sounds like we should make this change and make it configurable. Sound issues are the last thing we want in Atari800.

**Comment by nanoant** [comment]
I think the source of weird `33` silence value is `mkpokeysnd.c` turning `generate_sample()` `double` into `8-bit` value (let's skip `16-bit` case for now):
~~~c
buffer[0] = (UBYTE)floor((generate_sample(pokey_states) - MAX_SAMPLE / 2.0)
* (255.0 / MAX_SAMPLE / 4 * M_PI * 0.95) + 128 + 0.5 + 0.5 * rand() / RAND_MAX - 0.25);
~~~

It is easy to see the regardless of `MAX_SAMPLE`, feeding the formula above with zero gets us `33`!

Now the question is what is the value range of `generate_sample()` - is it signed `[-MAX_SAMPLE .. MAX_SAMPLE]` or unsigned `[0 .. MAX_SAMPLE]`?

If it is unsigned, why `generate_sample()` does produce `0` for silence, not `MAX_SAMPLE/2`?

If it is signed then the formula above is incorrect, it should be instead:
~~~c
buffer[0] = (UBYTE)floor(generate_sample(pokey_states)
* (128.0 / MAX_SAMPLE / 4 * M_PI * 0.95) + 128.0 + 0.5 + 0.5 * rand() / RAND_MAX - 0.25);
~~~

According to the *"Master gain and DC offset calculation"* description in the big comment block, it is more likely signed, talking about input range of `[-60 .. 60]` for resampler.

Thoughts?

**Comment by nanoant** [comment]
Sorry for hammering in there, but is there any chance to get this fixed? This DC bias humming is really annoying.

**Comment by mikrosk** [comment]
I'm absolutely no expert in this area but I wouldn't be surprised if the formula isn't correct. For instance I had to fix similar signed/unsigned problem when specifying certain compilation flags: https://github.com/atari800/atari800/commit/4024d4977aaf11c02c0af694bb1b2d7c2102bf48

**Comment by joysfera** [comment]
I'd like to get this issue fixed before releasing new Atari800 version. How can I help?
@nanoant have you tested your proposed formula by dumping the resulting WAV and looking at it?

**Comment by joysfera** [comment]
I'm looking into this. The generated samples (before the above mentioned conversion) are not signed/unsigned but rather "digital" with logical 0 and 1. Silence is generated as precise 0 while simple sound (say Pokey Self test sound of single channel) generates levels - the low is around 0 and the high is around 38 with slight oscillations (so the low can go below zero slightly).
Full four channel music then oscillates between -16 and 113.

The above mentioned conversion by @nanoant is actually working exactly as on real Atari computer (I have checked output of Atari 130 XE with oscilloscope, it looks just like the WAV from the emulator). This is great for the emulator accuracy but I can imagine a modern device might mis-interpret the silence that is not at zero level. I may try shifting it up a bit.

**Comment by joysfera** [comment]
Hopefully fixed in the 4.0.0 release. Thanks for reporting it.

**Comment by joysfera** [comment]
I wouldn't make it configurable unless the current (v4.0.0) setup creates any issue.
Atari computers had a lot of hardware shortcomings that we don't emulate either - like say PAL signal on video output that is not exactly compatible with the norm (and some modern TV sets have issues with that).
